### PR TITLE
LT-8656 Creating Running Flutter integration tests in Docker blog

### PR DESCRIPTION
**Pull request by phamnhuvu-dev**

### CODE REVIEW

**Review by vctqs1 — APPROVED** [review]

**Review by nploi — APPROVED** [review]

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+- Docker:
+    - An open platform for developing, shipping, and running applications
+    - https://docs.docker.com/get-started/overview/
+
+- Android Emulator container:
+    - Allowing you to find and run the right version of the Emulator without the headache of dependency management, which makes it easy to scale automated tests as part of a CI/CD system without the upkeep cost of a physical device farm
+    - https://github.com/google/android-emulator-container-scripts
+
+## The criteria that we use Docker for running automation tests
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: Why we use Docker to run automation tests

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+- Scaling tests:
+    - Docker provides the ability to package and run an application in a loosely isolated environment, so we can run multiple e2e tests parallel.
+
+- Reusing:
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md:39]: Reusability (?)

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    - Our Dockerfile file and docker-compose.yml file is simple because of reusing container from the community.
+
+- Consistent:
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: Consistency

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+COPY pubspec.* .
+
+RUN --mount=type=cache,sharing=locked,target=/flutter flutter pub get
+
+WORKDIR /project
+
+COPY . .
+```
+
+- To run the Flutter integration test we need a process for Android Emulator and a process for Flutter drive. That is why we have 2 services is `flutter-app` and `android-emulator` in `docker-compose.yml` file.
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - To run the Flutter integration test we need a process for Android Emulator and a process for Flutter drive. That is why we have 2 services: `flutter-app` and `android-emulator` in `docker-compose.yml` file.

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      - /dev/kvm
+
+  flutter-app:
+    depends_on:
+      - android-emulator
+    build: .
+    command: make run-test
+```
+
+- But `flutter-app` and `android-emulator` are services separately. It means they are not the same network, so we need to connect `flutter-app` to `android-emulator` by the following command: `adb connect android-emulator:5555`. Reference: https://developer.android.com/studio/command-line/adb#wireless
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - But `flutter-app` and `android-emulator` are 2 separate services . It means they are not the same network, so we need to connect `flutter-app` to `android-emulator` by the following command: `adb connect android-emulator:5555`. Reference: https://developer.android.com/studio/command-line/adb#wireless

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+- After `flutter-app` connects to `android-emulator`, we can run the Flutter integration tests with `flutter drive` command. We combine connect emulator script and flutter drive script in `run-test` command in `Makefile`.
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - After `flutter-app` connected to `android-emulator`, we can run the Flutter integration tests with `flutter drive` command. We combine connect emulator script and flutter drive script in `run-test` command in `Makefile`.

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+- After cloning the source code, move to the root of the source code.
+- Run `make run-docker-test`
+- If you see the following log after running the above command. You succeeded.
+
+```
+All tests passed!
+```
+
+## Summary:
+- Using docker for automation tests is used widely in the software industry, not only Mobile field.
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - Using docker for automation tests is being used widely in the software industry, not only Mobile field.

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+- With Docker's ability, scaling tests are easier and faster, helping us detect breakings and bugs quickly to fix and maintain.
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - With Docker's ability, scaling tests is far easier and faster, helping us to detect breakings and bugs quickly to fix and maintain.

File: content/posts/running-flutter-integration-tests-in-docker/index.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+- Maybe you have a question that almost all mobile developers use macOS to work. How can run the automation tests on macOS before shipping the source code to the cloud?
```
Comment (duytran1993) [inline content/posts/running-flutter-integration-tests-in-docker/index.md]: - Maybe you have the same question as almost all mobile developers who're using macOS to work. How can we run the automation tests on macOS before shipping the source code to the cloud?

File: content/posts/running-flutter-integration-tests-in-docker/index.md (hunk 2 above)
Comment (phamnhuvu-dev) [reply content/posts/running-flutter-integration-tests-in-docker/index.md:39]: No

**Review by duytran1993 — APPROVED** [review]

**Pull request merged**